connman/src/connection.c

```diff
@@ -107,34 +107,40 @@ static struct gateway_data *lookup_gateway_data(struct gateway_config *config)
 	return NULL;
 }
 
-static struct gateway_data *find_vpn_gateway(int index, const char *gateway)
+static struct gateway_data *find_vpn_gateway(int index, const char *gateway,
+						enum connman_ipconfig_type type)
 {
 	GHashTableIter iter;
 	gpointer value, key;
 
 	if (!gateway)
 		return NULL;
 
+	DBG("index %d type %d gateway %s", index, type, gateway);
+
 	g_hash_table_iter_init(&iter, gateway_hash);
 
 	while (g_hash_table_iter_next(&iter, &key, &value)) {
 		struct gateway_data *data = value;
 
 		if (data->ipv4_gateway && data->index4 == index &&
 				g_str_equal(data->ipv4_gateway->gateway,
-					gateway))
+					gateway) &&
+				type == CONNMAN_IPCONFIG_TYPE_IPV4)
```

Comment:
If 'type' is not ipv4 those other tests are meaningless -> you might want to do the type test first / something,

Reply:
Dropped.